integrations/upgrade/js-config.test.ts

```diff
@@ -468,3 +642,218 @@ test(
+         * Tailwind CSS v3. To make sure your project looks the same, we've added
+         * these compatibility styles to your project.
+         *
+         * It is recommended to migrate your project, and delete this layer once you
+         * are done.
+         *
+         * - All elements with a 'border' class will default to 'currentColor' in
+         *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+         * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+         *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+         */
+        @layer base {
+          *,
+          ::after,
+          ::before,
+          ::backdrop,
+          ::file-selector-button {
+            border-color: var(--color-gray-200, currentColor);
+          }
+
+          input:where(:not([type='button'], [type='reset'], [type='submit'])),
+          select,
+          textarea {
+            border-width: 0;
+          }
+        }
+        "
+      `)
+    },
+  )
+
+  test(
+    'migrate border compatibility if a custom border color is used',
+    {
+      fs: {
+        'package.json': json`
+          {
+            "dependencies": {
+              "@tailwindcss/upgrade": "workspace:^"
+            }
+          }
+        `,
+        'tailwind.config.ts': ts`
+          import { type Config } from 'tailwindcss'
+
+          export default {
+            theme: {
+              extend: {
+                borderColor: ({ colors }) => ({
+                  DEFAULT: colors.blue[500],
+                }),
+              },
+            },
+          } satisfies Config
+        `,
+        'src/input.css': css`
+          @tailwind base;
+          @tailwind components;
+          @tailwind utilities;
+        `,
+      },
+    },
+    async ({ exec, fs }) => {
+      await exec('npx @tailwindcss/upgrade')
+
+      expect(await fs.dumpFiles('src/**/*.css')).toMatchInlineSnapshot(`
+        "
+        --- src/input.css ---
+        @import 'tailwindcss';
+
+        /**
+         * The default border styles in Tailwind CSS v4 changed compared to 
+         * Tailwind CSS v3. To make sure your project looks the same, we've added
+         * these compatibility styles to your project.
+         *
+         * It is recommended to migrate your project, and delete this layer once you
+         * are done.
+         *
+         * - All elements with a 'border' class will default to 'currentColor' in
+         *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+         * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+         *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+         */
+        @layer base {
+          *,
+          ::after,
+          ::before,
+          ::backdrop,
+          ::file-selector-button {
+            border-color: var(--border-color, var(--color-gray-200, currentColor));
```

Comment:
This now contains `var(--border-color)` because it existed in your own tailwind.config.js file. We could also always generate this instead of just `var(--color-gray-200, currentColor)`.

Reply:
@RobinMalfait IIRC in the call yesterday we discussed maybe inlining the value of `--border-color` and omitting it from the created CSS (since it doesn't do anything for the core utilities anyways). WDYT about that approach?

Reply:
Yeah not a bad idea. I started with resolving the value using `theme(…)` and since it generated the variable and used it looked nice because you can still update the default.

However, the moment you get rid of the backwards compatibility CSS, then the variable is useless and vice versa.

Let me inline it instead.